File: scripts/appos/setup_dify_short_drama_workflow.ts
import fs from 'node:fs';
import path from 'node:path';

type CookieJar = Map<string, string>;

const root = 'B:\\Cir\\CodexProjects';
const envPath = path.join(root, 'opc-local.env');
const credentialsPath = path.join(root, 'opc-local.credentials.txt');
const dslPath = path.resolve('docs/appos/dify-workflows/short_drama_cps_edit_planner.yml');

const readEnvFile = (filePath: string) => {
  const env = new Map<string, string>();
  if (!fs.existsSync(filePath)) return env;
  for (const line of fs.readFileSync(filePath, 'utf8').split(/\r?\n/)) {
    if (!line || line.trim().startsWith('#') || !line.includes('=')) continue;
    const [key, ...rest] = line.split('=');
    env.set(key.trim(), rest.join('=').trim());
  }
  return env;
};

const readLocalCredentials = () => {
  const text = fs.existsSync(credentialsPath) ? fs.readFileSync(credentialsPath, 'utf8') : '';
  const email = text.match(/DIFY_ADMIN_EMAIL=(.+)/)?.[1]?.trim() || process.env.DIFY_ADMIN_EMAIL;
  const password = text.match(/DIFY_ADMIN_PASSWORD=(.+)/)?.[1]?.trim() || process.env.DIFY_ADMIN_PASSWORD;
  if (!email || !password) throw new Error('Missing Dify admin credentials in local private credentials file');
  return { email, password };
};

const cookieHeader = (jar: CookieJar) =>
  [...jar.entries()].map(([key, value]) => `${key}=${value}`).join('; ');

const storeCookies = (jar: CookieJar, headers: Headers) => {
  const raw = (headers as Headers & { getSetCookie?: () => string[] }).getSetCookie?.() ?? [];
  const cookies = raw.length ? raw : headers.get('set-cookie')?.split(/,(?=\s*[^;,]+=)/) ?? [];
  for (const cookie of cookies) {
    const pair = cookie.split(';')[0];
    const index = pair.indexOf('=');
    if (index > 0) jar.set(pair.slice(0, index).trim(), pair.slice(index + 1).trim());
  }
};

const requestJson = async (url: string, options: RequestInit & { jar: CookieJar }) => {
  const headers = new Headers(options.headers);
  if (options.jar.size > 0) headers.set('cookie', cookieHeader(options.jar));
  const response = await fetch(url, { ...options, headers });
  storeCookies(options.jar, response.headers);
  const text = await response.text();
  let data: unknown = text;
  try {
    data = text ? JSON.parse(text) : {};
  } catch {
    // keep text
  }
  if (!response.ok) {
    throw new Error(`Dify API ${response.status} ${url}: ${typeof data === 'string' ? data.slice(0, 500) : JSON.stringify(data).slice(0, 500)}`);
  }
  return data as Record<string, unknown>;
};

const upsertEnvLine = (filePath: string, key: string, value: string) => {
  const lines = fs.existsSync(filePath) ? fs.readFileSync(filePath, 'utf8').split(/\r?\n/) : [];
  const prefix = `${key}=`;
  let replaced = false;
  const next = lines.map((line) => {
    if (line.startsWith(prefix)) {
      replaced = true;
      return `${key}=${value}`;
    }
    return line;
  });
  if (!replaced) next.push(`${key}=${value}`);
  fs.writeFileSync(filePath, `${next.filter((line, index) => line || index < next.length - 1).join('\n')}\n`, 'utf8');
};

async function main() {
  const env = readEnvFile(envPath);
  const apiBaseUrl = (env.get('DIFY_API_URL') || process.env.DIFY_API_URL || 'http://127.0.0.1:5001').replace(/\/$/, '');
  const { email, password } = readLocalCredentials();
  const jar: CookieJar = new Map();

  await requestJson(`${apiBaseUrl}/console/api/login`, {
    jar,
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({ email, password, remember_me: true })
  });

  const yamlContent = fs.readFileSync(dslPath, 'utf8');
  const imported = await requestJson(`${apiBaseUrl}/console/api/apps/imports`, {
    jar,
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({
      mode: 'yaml-content',
      yaml_content: yamlContent,
      name: 'OPC 短剧CPS剪辑策划',
      description: '短剧 CPS 第 6 步 Dify 剪辑策划 workflow'
    })
  });

  const appId = String(imported.app_id || '');
  if (!appId) throw new Error(`Dify import did not return app_id: ${JSON.stringify(imported)}`);

  await requestJson(`${apiBaseUrl}/console/api/apps/${appId}/workflows/publish`, {
    jar,
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({ marked_name: 'v1', marked_comment: 'Tele-OPC short drama CPS edit planner' })
  });

  const apiKey = await requestJson(`${apiBaseUrl}/console/api/apps/${appId}/api-keys`, {
    jar,
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({})
  });

  const token = String(apiKey.token || '');
  if (!token) throw new Error(`Dify API key creation did not return token: ${JSON.stringify(apiKey)}`);

  upsertEnvLine(envPath, 'APPOS_DIFY_SHORT_DRAMA_CPS_WORKFLOW_URL', `${apiBaseUrl}/v1/workflows/run`);
  upsertEnvLine(envPath, 'APPOS_DIFY_SHORT_DRAMA_CPS_APP_ID', appId);
  upsertEnvLine(envPath, 'APPOS_DIFY_SHORT_DRAMA_CPS_API_KEY', token);

  console.log(JSON.stringify({ ok: true, appId, workflowUrl: `${apiBaseUrl}/v1/workflows/run`, apiKeyStoredIn: envPath }, null, 2));
}

main().catch((error) => {
  console.error(error instanceof Error ? error.message : String(error));
  process.exitCode = 1;
});

File: scripts/appos/setup_dify_short_drama_workflow.py
from __future__ import annotations

import json
import os
import re
import sys
from pathlib import Path

from sqlalchemy import select
from sqlalchemy.orm import Session


ROOT = Path("B:/Cir/CodexProjects")
ENV_PATH = ROOT / "opc-local.env"
CREDENTIALS_PATH = ROOT / "opc-local.credentials.txt"
DSL_PATH = Path("B:/Cir/CodexProjects/tele-opc/docs/appos/dify-workflows/short_drama_cps_edit_planner.yml")
DIFY_API_ROOT = ROOT / "dify" / "api"
DIFY_ENV_PATH = DIFY_API_ROOT / ".env"
DEFAULT_LOG_FORMAT = "%(asctime)s,%(msecs)d %(levelname)-2s [%(filename)s:%(lineno)d] %(req_id)s %(message)s"


def read_env_file(path: Path) -> dict[str, str]:
    values: dict[str, str] = {}
    if not path.exists():
        return values
    for line in path.read_text(encoding="utf-8").splitlines():
        if not line or line.lstrip().startswith("#") or "=" not in line:
            continue
        key, value = line.split("=", 1)
        values[key.strip()] = value.strip()
    return values


def read_admin_email() -> str:
    text = CREDENTIALS_PATH.read_text(encoding="utf-8") if CREDENTIALS_PATH.exists() else ""
    email = re.search(r"DIFY_ADMIN_EMAIL=(.+)", text)
    value = (email.group(1).strip() if email else os.getenv("DIFY_ADMIN_EMAIL", "")).strip()
    if not value:
        raise RuntimeError("Missing DIFY_ADMIN_EMAIL in local private credentials file")
    return value


def upsert_env_line(path: Path, key: str, value: str) -> None:
    lines = path.read_text(encoding="utf-8").splitlines() if path.exists() else []
    prefix = f"{key}="
    replaced = False
    next_lines: list[str] = []
    for line in lines:
        if line.startswith(prefix):
            next_lines.append(f"{key}={value}")
            replaced = True
        else:
            next_lines.append(line)
    if not replaced:
        next_lines.append(f"{key}={value}")
    path.write_text("\n".join(next_lines).rstrip() + "\n", encoding="utf-8")


def apply_env_defaults(*env_files: Path) -> None:
    for path in env_files:
        for key, value in read_env_file(path).items():
            os.environ.setdefault(key, value)
    os.environ["LOG_OUTPUT_FORMAT"] = os.environ.get("LOG_OUTPUT_FORMAT") or "text"
    if os.environ.get("LOG_OUTPUT_FORMAT") == "json":
        os.environ["LOG_OUTPUT_FORMAT"] = "text"
    if os.environ.get("LOG_FORMAT") in {"", "json", None}:
        os.environ["LOG_FORMAT"] = DEFAULT_LOG_FORMAT


def main() -> None:
    apply_env_defaults(ENV_PATH, DIFY_ENV_PATH)
    if str(DIFY_API_ROOT) not in sys.path:
        sys.path.insert(0, str(DIFY_API_ROOT))

    env = read_env_file(ENV_PATH)
    api_base_url = env.get("DIFY_API_URL") or os.getenv("DIFY_API_URL") or "http://127.0.0.1:5001"
    email = read_admin_email().lower()
    yaml_content = DSL_PATH.read_text(encoding="utf-8")

    from app_factory import create_app
    from extensions.ext_database import db
    from libs.datetime_utils import naive_utc_now
    from models.account import Account
    from models.enums import ApiTokenType
    from models.model import ApiToken, App
    from services.account_service import TenantService
    from services.app_dsl_service import AppDslService
    from services.entities.dsl_entities import ImportMode, ImportStatus
    from services.workflow_service import WorkflowService

    _socketio_app, flask_app = create_app()
    with flask_app.app_context():
      with Session(db.engine, expire_on_commit=False) as session:
        account = session.scalar(select(Account).where(Account.email == email))
        if account is None:
            raise RuntimeError(f"Dify account not found for configured admin email: {email}")
        joins = TenantService.get_join_tenants(account, session=session)
        if not joins:
            raise RuntimeError(f"Dify account has no workspace tenant: {email}")
        account.current_tenant = joins[0]
        account.set_tenant_id(joins[0].id)

        import_service = AppDslService(session)
        result = import_service.import_app(
            account=account,
            import_mode=ImportMode.YAML_CONTENT,
            yaml_content=yaml_content,
            name="OPC 短剧CPS剪辑策划",
            description="短剧 CPS 第 6 步 Dify 剪辑策划 workflow",
            icon_type="emoji",
            icon="🎬",
            icon_background="#FFEAD5",
        )
        if result.status not in {ImportStatus.COMPLETED, ImportStatus.COMPLETED_WITH_WARNINGS}:
            raise RuntimeError(f"Dify import failed: {result.model_dump(mode='json')}")
        if not result.app_id:
            raise RuntimeError(f"Dify import returned no app_id: {result.model_dump(mode='json')}")

        app_model = session.get(App, result.app_id)
        if app_model is None:
            raise RuntimeError(f"Dify app not found after import: {result.app_id}")

        workflow = WorkflowService().publish_workflow(
            session=session,
            app_model=app_model,
            account=account,
            marked_name="v1",
            marked_comment="Tele-OPC short drama CPS edit planner",
        )
        app_model.workflow_id = workflow.id
        app_model.updated_by = account.id
        app_model.updated_at = naive_utc_now()

        api_token = ApiToken()
        api_token.app_id = app_model.id
        api_token.tenant_id = app_model.tenant_id
        api_token.token = ApiToken.generate_api_key("app-", 24)
        api_token.type = ApiTokenType.APP
        session.add(api_token)
        session.commit()

        upsert_env_line(ENV_PATH, "APPOS_DIFY_SHORT_DRAMA_CPS_WORKFLOW_URL", f"{api_base_url.rstrip('/')}/v1/workflows/run")
        upsert_env_line(ENV_PATH, "APPOS_DIFY_SHORT_DRAMA_CPS_APP_ID", str(app_model.id))
        upsert_env_line(ENV_PATH, "APPOS_DIFY_SHORT_DRAMA_CPS_API_KEY", api_token.token)

        print(json.dumps({
            "ok": True,
            "appId": str(app_model.id),
            "workflowId": str(workflow.id),
            "workflowUrl": f"{api_base_url.rstrip('/')}/v1/workflows/run",
            "apiKeyStoredIn": str(ENV_PATH),
        }, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()
